src/rust/engine/options/src/config.rs

```diff
@@ -1,18 +1,98 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::mem;
 use std::path::Path;
 
+use lazy_static::lazy_static;
+use regex::Regex;
 use toml::value::Table;
 use toml::Value;
 
 use super::id::{NameTransform, OptionId};
 use super::parse::parse_string_list;
 use super::{ListEdit, ListEditAction, OptionsSource};
 
+type InterpolationMap = HashMap<String, String>;
+
+lazy_static! {
+    static ref PLACEHOLDER_RE: Regex = Regex::new(r"%\(([a-zA-Z0-9_.]+)\)s").unwrap();
+}
+
+pub(crate) fn interpolate_string(
+    value: String,
+    replacements: &InterpolationMap,
+) -> Result<String, String> {
```

Comment:
So... I don't think that you really need to worry about it here, but: this is precisely the situation that `Cow` was designed for: "a value might be either an owned, modified value OR a borrowed, unmodified value".

If you are interested in learning about `Cow` and using it in your return type here, then: https://dhghomon.github.io/easy_rust/Chapter_42.html ... otherwise, please just leave a TODO comment here suggesting that this should use `Cow`.

Reply:
I did originally implement this with Cow, but I realized that it was unnecessary in this case, since we can always consume, and need not borrow, the value String: We are destructively iterating over the uninterpolated config, so we take ownership of all its strings, and so there is no more allocation happening in this implementation than there would be if we used Cow. And since it made the signature (and the recursion, in particular) more complicated, I switched away from Cow.